Mathlib/LinearAlgebra/Matrix/Adjugate.lean

```diff
@@ -0,0 +1,541 @@
+/-
+Copyright (c) 2019 Anne Baanen. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Anne Baanen
+
+! This file was ported from Lean 3 source module linear_algebra.matrix.adjugate
+! leanprover-community/mathlib commit 0e2aab2b0d521f060f62a14d2cf2e2c54e8491d6
+! Please do not edit these lines, except to modify the commit id
+! if you have ported upstream changes.
+-/
+import Mathlib.Algebra.Regular.Basic
+import Mathlib.LinearAlgebra.Matrix.MvPolynomial
+import Mathlib.LinearAlgebra.Matrix.Polynomial
+import Mathlib.RingTheory.Polynomial.Basic
+
+/-!
+# Cramer's rule and adjugate matrices
+
+The adjugate matrix is the transpose of the cofactor matrix.
+It is calculated with Cramer's rule, which we introduce first.
+The vectors returned by Cramer's rule are given by the linear map `cramer`,
+which sends a matrix `A` and vector `b` to the vector consisting of the
+determinant of replacing the `i`th column of `A` with `b` at index `i`
+(written as `(A.update_column i b).det`).
+Using Cramer's rule, we can compute for each matrix `A` the matrix `adjugate A`.
+The entries of the adjugate are the minors of `A`.
+Instead of defining a minor by deleting row `i` and column `j` of `A`, we
+replace the `i`th row of `A` with the `j`th basis vector; the resulting matrix
+has the same determinant but more importantly equals Cramer's rule applied
+to `A` and the `j`th basis vector, simplifying the subsequent proofs.
+We prove the adjugate behaves like `det A • A⁻¹`.
+
+## Main definitions
+
+ * `Matrix.cramer A b`: the vector output by Cramer's rule on `A` and `b`.
+ * `Matrix.adjugate A`: the adjugate (or classical adjoint) of the matrix `A`.
+
+## References
+
+  * https://en.wikipedia.org/wiki/Cramer's_rule#Finding_inverse_matrix
+
+## Tags
+
+cramer, cramer's rule, adjugate
+-/
+
+
+-- Porting note: needed to make `cramer` function application work
+set_option synthInstance.etaExperiment true
+
+namespace Matrix
+
+universe u v w
+
+variable {m : Type u} {n : Type v} {α : Type w}
+
+variable [DecidableEq n] [Fintype n] [DecidableEq m] [Fintype m] [CommRing α]
+
+open Matrix BigOperators Polynomial Equiv Equiv.Perm Finset
+
+section Cramer
+
+/-!
+  ### `cramer` section
+
+  Introduce the linear map `cramer` with values defined by `cramer_map`.
+  After defining `cramer_map` and showing it is linear,
+  we will restrict our proofs to using `cramer`.
+-/
+
+
+variable (A : Matrix n n α) (b : n → α)
+
+/-- `cramer_map A b i` is the determinant of the matrix `A` with column `i` replaced with `b`,
+  and thus `cramer_map A b` is the vector output by Cramer's rule on `A` and `b`.
+
+  If `A ⬝ x = b` has a unique solution in `x`, `cramer_map A` sends the vector `b` to `A.det • x`.
+  Otherwise, the outcome of `cramer_map` is well-defined but not necessarily useful.
```

Comment:
```suggestion
/-- `cramerMap A b i` is the determinant of the matrix `A` with column `i` replaced with `b`,
  and thus `cramerMap A b` is the vector output by Cramer's rule on `A` and `b`.

  If `A ⬝ x = b` has a unique solution in `x`, `cramerMap A` sends the vector `b` to `A.det • x`.
  Otherwise, the outcome of `cramerMap` is well-defined but not necessarily useful.
```